src/thermo/ConstDensityTabulatedThermo.cpp

```diff
@@ -0,0 +1,203 @@
+		m_spthermo.modifySpecies(m_kk_mod, stit);
+		m_spthermo.update(tnow, &m_cp0_R[0], &m_h0_RT[0],
+				&m_s0_R[0]);
+		for (size_t k = 0; k < m_kk; k++) {
+			m_g0_RT[k] = m_h0_RT[k] - m_s0_R[k];
+		}
+		m_xlast = xnow;
+		m_tlast = tnow;
+	}
+}
+
+void ConstDensityTabulatedThermo::initThermoXML(XML_Node& phaseNode, const std::string& id_)
+{
+	ConstDensityTabulatedThermo::initThermo();
+	size_t offset;
+	doublereal a=0,b=0,c=0;
+	std::ifstream t;
+	std::string line;
+	int count = 0;
+
+	if (phaseNode.hasChild("thermo")) {
+		XML_Node& thermoNode = phaseNode.child("thermo");
+		std::string model = thermoNode["model"];
+		if (model == "ConstDensityTabulatedThermo") {
+			if (thermoNode.hasChild("data")) {
+				XML_Node& scNode = thermoNode.child("data");
+				std::vector<std::string> nameFile;
+				getStringArray(scNode, nameFile);
+				if (nameFile.size() != 1) {
+					throw CanteraError("ConstDensityTabulatedThermo::initThermoXML",
+							"badly formed data XML node");
+				}
+				m_dataFile = nameFile[0];
+			} else {
+				throw CanteraError("initThermoXML",
+						"Unspecified data file");
+			}
+			if (thermoNode.hasChild("modifiable_species")) {
+				std::string modifiable_species_name = thermoNode.child("modifiable_species").value();
+				m_kk_mod = speciesIndex(modifiable_species_name);
+				if (m_kk_mod == npos) {
+					throw CanteraError("ConstDensityTabulatedThermo::initThermoXML",
+							"Species " + modifiable_species_name + " not found.");
+				}
+			} else {
+				throw CanteraError("ConstDensityTabulatedThermo::initThermoXML",
+						"Unspecified modifiable species");
+			}
+			m_xlast = moleFraction(m_kk_mod);
+		} else {
+			throw CanteraError("ConstDensityTabulatedThermo::initThermoXML",
+					"Unknown thermo model : " + model);
+		}
+	}
+	t.open(m_dataFile.c_str());
+	if(t.fail())
+	{
+		throw CanteraError("initThermoXML",
+				"Unknown data file: " + m_dataFile);
+	}
+	while (getline(t,line))
+	{
+		if((offset = line.find("*", 0)) != std::string::npos)
+		{
+			continue;
+		}
+		else
+		{
+			count = std::sscanf(line.c_str(), "%lf %lf %lf", &a, &b, &c);
+			if (count==3)
+			{
+				molefrac_h.push_back(std::make_pair(a,b));
+				molefrac_s.push_back(std::make_pair(a,c));
+			}
+			else
+			{
+				throw CanteraError("initThermoXML",
+						"Data format error: " + m_dataFile);
+			}
+		}
+	}
+	// Sort the xLi, h, s data in the order of increasing xLi
+	std::sort(molefrac_h.begin(), molefrac_h.end());
+	std::sort(molefrac_s.begin(), molefrac_s.end());
+	t.close();
+	ThermoPhase::initThermoXML(phaseNode, id_);
+}
+
+doublereal ConstDensityTabulatedThermo::interp_h(doublereal x) const
+{
+	// Assumes that "table" is sorted by .first
+	// Check if x is out of bound
+	if (x > molefrac_h.back().first) return molefrac_h.back().second;
+	if (x < molefrac_h[0].first) return molefrac_h[0].second;
+	std::vector<std::pair<double, double> >::const_iterator it, it2;
+	// INFINITY is defined in math.h in the glibc implementation

```

Comment:
I don't see how this comment is related to the adjacent lines of code. Can you explain a bit more please?